src/machine.rs

```diff
@@ -81,6 +84,7 @@ pub struct AllocExtra {
 pub struct MemoryExtra {
     pub stacked_borrows: Option<stacked_borrows::MemoryExtra>,
     pub intptrcast: intptrcast::MemoryExtra,
+    pub tls: ThreadLocalStorage,
```

Comment:
It seems really strange to have *both* this and the `tls` field in `Evaluator`. Would it make sense to merge them? At the very least there should be an explanation for why they are separate.

Reply:
Oh I guess one is for "native TLS" (`#[thread_local]`) and the other for OS-library-supported TLS. It is quite annoying that both exist. :/

Can we somehow give them clear names to distinguish the two?

Reply:
> Oh I guess one is for "native TLS" (#[thread_local]) and the other for OS-library-supported TLS. 

That is exactly the reason. I do not think it would make sense to merge them because their implementations have nothing in common.

> Can we somehow give them clear names to distinguish the two?

Maybe `lang_tls` and `library_tls`?

Reply:
> Maybe lang_tls and library_tls?

Sounds like a good start. :)

Reply:
The new callback in Rustc gives me access to the whole interpreter context, which makes this field unnecessary.